lib/bolt/inventory/inventory2.rb

```diff
@@ -219,102 +197,202 @@ def expand_targets(targets)
           targets.split(/[[:space:],]+/).reject(&:empty?).map do |name|
             ts = resolve_name(name)
             ts.map do |t|
-              target = create_target(t)
-              target.inventory = self
-              target
+              # If the target exists, return it, otherwise create one
+              @targets[t] ? @targets[t]['self'] : create_target(t)
             end
           end
         end
       end
       private :expand_targets
 
-      def set_vars_from_hash(tname, data)
-        if data
-          # Instantiate empty vars hash in case no vars are defined
-          @target_vars[tname] ||= {}
-          # Assign target new merged vars hash
-          # This is essentially a copy-on-write to maintain the immutability of @target_vars
-          @target_vars[tname] = @target_vars[tname].merge(data).freeze
+      def ensure_in_all_group(targets)
+        names = flat_target_list(targets).map(&:name)
+        targets_to_add = Set.new(names) - @groups.target_names
+        targets_to_add.each { |t| add_to_group([@targets[t]['self']], 'all') }
+      end
+
+      def add_target(current_group, target, desired_group)
+        if current_group.name == desired_group
+          current_group.add_target(target.target_data_hash)
+          @groups.validate
+          update_target(target)
+          return true
+        end
+        # Recurse on children Groups if not desired_group
+        current_group.groups.each do |child_group|
+          add_target(child_group, target, desired_group)
         end
       end
-      private :set_vars_from_hash
+      private :add_target
+
+      # CODEREVEW: For now these methods live in inventory, should it be a subclass or some other module?
```

Comment:
Since these are both reading from and writing to state on the `Inventory` class it seems like managing another class or module would only complicate things (either having to pass data back and forth, or making `@groups` readable and `@targets` writable, plus manage the new class' lifecycle). It makes sense (to me at least?) for these methods to be operating on the Inventory.

Reply:
@puppetlabs/bolt Open question